### PR TITLE
feat: support multiple GatewayClass per controller

### PR DESCRIPTION
**Pull request by cnvergence**
**What type of PR is this?**

**What this PR does / why we need it**:

- supporting multiple GatewayClass with different infrastructure deployment types
- extend Provider Resources watchable interface with GatewayClassResources map
- initial e2e test for multiple GatewayClass scenario

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/1231

### CODE REVIEW

**Comment by arkodg** [comment]
@cnvergence kinda amazing to see how this PR doesnt require any changes to the gateway-api layer :)

File: internal/provider/kubernetes/controller.go [hunk 1]
```diff
@@ -180,213 +181,206 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, _ reconcile.Reques
...
+			if err := r.addFinalizer(ctx, acceptedGC); err != nil {
+				r.log.Error(err, fmt.Sprintf("failed adding finalizer to gatewayclass %s",
+					acceptedGC.Name))
+				return reconcile.Result{}, err
+			}
 		}
+		// The Store is triggered even when there are no Gateways associated to the
+		// GatewayClass. This would happen in case the last Gateway is removed and the
+		// Store will be required to trigger a cleanup of envoy infra resources.
+		r.resources.GatewayAPIResources.Store(acceptedGC.Name, resourcesMap[acceptedGC.Name])
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:381]: seeing a bug 
* the gateway-api translator works at a few gateway class level 
https://github.com/envoyproxy/gateway/blob/e168ef961dbd40861af401bf70c781e2cba282fd/internal/gatewayapi/runner/runner.go#L63
* however in the gateway-api layer, the logic to clean up IRs is global, which will cause other gateways to be deleted
https://github.com/envoyproxy/gateway/blob/e168ef961dbd40861af401bf70c781e2cba282fd/internal/gatewayapi/runner/runner.go#L172

Possible solution
* filter on relevant keys (try and match GC) when cleaning
* or keep it flat, and pass 1 provider resource to the gateway-api layer with a list of GCs (the gateway-api translator can still work on 1 GC at a time)

Comment (cnvergence) [reply internal/provider/kubernetes/controller.go:381]: right I checked it out, thanks @arkodg

File: internal/gatewayapi/runner/runner.go [hunk 2]
```diff
@@ -76,10 +77,12 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 			}
 			// Translate to IR
 			result := t.Translate(val)
-
 			var curKeys, newKeys []string
 			// Get current IR keys
-			for key := range r.InfraIR.LoadAll() {
+			filteredInfra := r.InfraIR.LoadAllMatching(func(_ string, ir *ir.Infra) bool {
```
Comment (arkodg) [inline internal/gatewayapi/runner/runner.go:82]: im a little wary of relying on labels to decide translation.here is one suggestion to solve this, please lmk what you think
* change the pub-sub watchable interface to publish a `providerResources` which looks like `map[string]*gatewayi.Resources` where the key is the gateway class name
* loop through the `providerResources` and for each gatewayClass, call `gatewayapi.Translate`
* remember all the IR values
* once done looping, clean up any old IRs
* publish all IRs

Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:82]: Understood, I thought that it would be safe anyway to keep labels as they are used in IR only.
The problem was with dealing with the keys, runner already watches for changes in map[string]*gatewayi.Resources, where the key is a GatewayClass name and InfraIR has a key in the Gateway namespace/ Gateway name format or GatewayClass Name for merged gateways.

Comment (arkodg) [reply internal/gatewayapi/runner/runner.go:82]: not sure I understood the problem, can you rephrase that

Comment (arkodg) [reply internal/gatewayapi/runner/runner.go:82]: I understand your issue :) , the watchable message needs to be a map representing all the resources, so the key here needs to the controller name, but its value is also a map which is a list of tuples with key as gatewayclass name and val as child resources

Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:82]: sorry, it was easy to get confused, I should rephrased that it wasn't that big of an issue, just the label approach was much simpler than matching all gateways from Infra keys :)

Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:82]: ok, so you would like to change the interface completely and rework this update func, now I think I understand your idea
let me try this and I will get back with it

Comment (arkodg) [reply internal/gatewayapi/runner/runner.go:82]: thanks @cnvergence !

Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:82]: hey @arkodg, you would like to have this watchable message look like?
```code
type ControlledResources struct {
	ControlledResources watchable.Map[string, *ProviderResources]
}

// ProviderResources message
type ProviderResources struct {
	// GatewayAPIResources is a map from a GatewayClass name to
	// a group of gateway API and other related resources.
	GatewayAPIResources watchable.Map[string, *gatewayapi.Resources]

	// GatewayAPIStatuses is a group of gateway api
	// resource statuses maps.
	GatewayAPIStatuses

	// PolicyStatuses is a group of policy statuses maps.
	PolicyStatuses
}
```
this seems to be a big refactor, so want to make sure

Comment (arkodg) [reply internal/gatewayapi/runner/runner.go:82]:
```
type ControllerResources map[string]*gatewayapi.Resources

// ProviderResources message
type ProviderResources struct {
	// GatewayAPIResources is a map from a GatewayClass name to
	// a group of gateway API and other related resources.
	GatewayAPIResources watchable.Map[string, *ControllerResources]

	// GatewayAPIStatuses is a group of gateway api
	// resource statuses maps.
	GatewayAPIStatuses

	// PolicyStatuses is a group of policy statuses maps.
	PolicyStatuses
}
```

probably something more like this, wdyt ?

Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:82]: That should do the work for the translator and matching old IR, I will check on how to use the new ProviderResources in the kubernetes provider.

File: internal/gatewayapi/resource.go [hunk 3]
```diff
@@ -20,6 +20,37 @@ import (
 
 type XdsIRMap map[string]*ir.Xds
 type InfraIRMap map[string]*ir.Infra
+type GatewayClassResources map[string]*Resources
+
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
Comment (arkodg) [inline internal/gatewayapi/resource.go:25]: did you hand write this ?
you can use a annotation similar to https://github.com/envoyproxy/gateway/blob/988d4ed0be5a5f9dd39a784d52c3bb5c51ecc2aa/internal/gatewayapi/resource.go#L26 to autogenerate it

File: internal/gatewayapi/runner/runner.go [hunk 4]
```diff
@@ -49,118 +49,117 @@ func (r *Runner) Start(ctx context.Context) (err error) {
...
-			// Also validate the ir before sending it.
-			for key, val := range result.InfraIR {
-				r.Logger.WithValues("infra-ir", key).Info(val.YAMLString())
-				if err := val.Validate(); err != nil {
-					r.Logger.Error(err, "unable to validate infra ir, skipped sending it")
-					errChan <- err
-				} else {
-					r.InfraIR.Store(key, val)
-					newKeys = append(newKeys, key)
+			for gc, resources := range *val {
```
Comment (arkodg) [inline internal/gatewayapi/runner/runner.go:65]: lets protect against `val == nil`

File: internal/gatewayapi/resource.go [hunk 5]
```diff
@@ -20,6 +20,37 @@ import (
 
 type XdsIRMap map[string]*ir.Xds
 type InfraIRMap map[string]*ir.Infra
+type GatewayClassResources map[string]*Resources
```
Comment (arkodg) [inline internal/gatewayapi/resource.go:23]: non blocking comment that can be tackled in a new PR - would a list `*Resources[]` be a better option here ?

File: internal/provider/kubernetes/controller.go [hunk 6]
```diff
@@ -165,212 +166,206 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, _ reconcile.Reques
...
-				acceptedGC.Name))
-			return reconcile.Result{}, err
-		}
-	} else {
-		// finalize the accepted GatewayClass.
-		if err := r.addFinalizer(ctx, acceptedGC); err != nil {
-			r.log.Error(err, fmt.Sprintf("failed adding finalizer to gatewayclass %s",
-				acceptedGC.Name))
-			return reconcile.Result{}, err
+			// If needed, remove the finalizer from the accepted GatewayClass.
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:353]: looks like the order is changed here a bit, shouldnt we update the finalizer and then run `Store`, so any labels show up in the gatewayapi layer and can be used/if required in the future ?

**Comment by arkodg** [comment]
minor comments, but this is looking good @cnvergence ! massive feature, that im hoping lands in v1.
0.0

File: internal/gatewayapi/runner/runner.go (hunk 4 above)
Comment (cnvergence) [reply internal/gatewayapi/runner/runner.go:65]: checking before dereferencing
```
val := update.Value
if update.Delete || val == nil {
	return
}
```

File: internal/gatewayapi/resource.go (hunk 3 above)
Comment (cnvergence) [reply internal/gatewayapi/resource.go:25]: I did generate it, I will add comment
deepcopy-gen returns GatewayClassResources, I needed to change it to return *GatewayClassResources

File: internal/provider/kubernetes/controller.go (hunk 6 above)
Comment (cnvergence) [reply internal/provider/kubernetes/controller.go:353]: I have changed the order, so the store was called when the last gateway was removed and then clean up finalizer. 
I could verify that running the store after the finalizers sometimes resulted in not cleaning up infra resources, e.g. deleting all resources at once.

File: internal/gatewayapi/resource.go (hunk 3 above)
Comment (arkodg) [reply internal/gatewayapi/resource.go:25]: because its a `map` ? , if it was a list would we hit this same issue ?

**Comment by arkodg** [comment]
@cnvergence looks like e2e is failing

File: internal/provider/kubernetes/controller.go (hunk 6 above)
Comment (cnvergence) [reply internal/provider/kubernetes/controller.go:350]: I was wrong, now with the new structure, it should be called afterwards

**Comment by cnvergence** [comment]
fixed those issues

File: internal/gatewayapi/resource.go (hunk 3 above)
Comment (cnvergence) [reply internal/gatewayapi/resource.go:25]: yes, I think it is because of the map

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Review by shawnh2 — APPROVED** [review]
LGTM, thanks
